core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -131,68 +128,72 @@
    * @param otherFields other fields to index the set
    */
   public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    mIndexMap = new HashMap<FieldIndex<T>, Integer>(otherFields.length + 1);
-    mIndexMap.put(field, 0);
+    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
+        new HashMap<>(otherFields.length + 1);
+    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>());
     for (int i = 1; i <= otherFields.length; i++) {
-      mIndexMap.put(otherFields[i - 1], i);
-    }
-
-    mSetIndexedByFieldValue = new ArrayList<Map<Object, Set<T>>>(mIndexMap.size());
-    for (int i = 0; i < mIndexMap.size(); i++) {
-      mSetIndexedByFieldValue.add(new HashMap<Object, Set<T>>());
+      indexMap.put(otherFields[i - 1], new ConcurrentHashMap<Object, ConcurrentHashSet<T>>());
     }
+    // read only, so it is thread safe and allows concurrent access.
+    mIndexMap = Collections.unmodifiableMap(indexMap);
   }
 
   /**
    * Removes all the entries in this set.
    */
+  // TODO(gpang): clear() does not play well with other concurrent operations. Currently, it is
```

Comment:
I just made clear() call delete() on each element. That does make it more expensive, so I left a comment.

This is actually similar behavior to ConcurrentHashMap.clear() since that also allows concurrent put().